plugins/rest/rest_auth.go

```diff
@@ -122,6 +104,112 @@ func (ap *bearerAuthPlugin) Prepare(req *http.Request) error {
 	return nil
 }
 
+type tokenEndpointResponse struct {
+	AccessToken string `json:"access_token"`
+	TokenType   string `json:"token_type"`
+	ExpiresIn   int64  `json:"expires_in"`
+}
+
+// oauth2ClientCredentialsAuthPlugin represents authentication via a bearer token in the HTTP Authorization header
+// obtained through the OAuth2 client credentials flow
+type oauth2ClientCredentialsAuthPlugin struct {
+	TokenURL     string    `json:"token_url"`
+	ClientID     string    `json:"client_id"`
+	ClientSecret string    `json:"client_secret"`
+	Scopes       *[]string `json:"scopes,omitempty"`
+
+	tokenCache    *oauth2Token
+	tlsSkipVerify bool
+}
+
+type oauth2Token struct {
+	Token     string
+	ExpiresAt time.Time
+}
+
+func (ap *oauth2ClientCredentialsAuthPlugin) NewClient(c Config) (*http.Client, error) {
+	t, err := defaultTLSConfig(c)
+	if err != nil {
+		return nil, err
+	}
+
+	// Inherit skip verify from the "parent" settings. Should this be configurable on the credentials too?
+	ap.tlsSkipVerify = c.AllowInsureTLS
```

Comment:
Right, this gets tricky because there is a need for essentially all of these cert/header settings for the client used to make these requests.

🤔   We almost want to use another service definition for the oauth client

Reply:
Yeah, I mainly tried to avoid an explosion of config options here but I agree they could be useful and might be needed later. Another one we'll likely want to consider at some point is the option to pass in arbitrary parameters in the token request. I know auth0 requires a non-standard "audience" parameter to be present, and there are most likely other implementations that have their own quirks, but I figured I would just aim for a good baseline first and then we can refine this based off of actual use and user requirements later. A separate service definition sounds like an interesting option if this grows bigger over time.